gobblin-compaction/src/main/java/gobblin/compaction/mapreduce/MRCompactor.java

```diff
@@ -522,15 +530,82 @@ private void verifyDataCompleteness() {
    */
   private boolean shouldVerifyCompletenessForDataset(Dataset dataset, List<Pattern> blacklist,
       List<Pattern> whitelist) {
-    return !datasetAlreadyCompacted(this.fs, dataset)
+      boolean renameDir = this.state.getPropAsBoolean(COMPACTION_RENAME_SOURCE_DIR, DEFAULT_COMPACTION_RENAME_SOURCE_DIR);
+      if (renameDir) {
+          LOG.info ("Should verify completeness with rename #### " + renameDir);
+      }
+    return !datasetAlreadyCompacted(this.fs, dataset, renameDir)
         && DatasetFilterUtils.survived(dataset.getName(), blacklist, whitelist);
   }
 
   /**
-   * A {@link Dataset} is considered already compacted if there is a file named
-   * {@link MRCompactor#COMPACTION_COMPLETE_FILE_NAME} in its {@link Dataset#outputPath()}.
+   * Get all the renamed directories from {@link Dataset#inputPaths}
+   * They are deepest level directories which have a suffix
+   * {@link MRCompactor#COMPACTION_RENAME_SOURCE_DIR_SUFFIX} append.
+   */
+  public static Set<Path> getInputDeepestLevelRenamedDirsWithileExistence (FileSystem fs, Set<Path> paths) throws IOException {
+    Set<Path> renamedDirs = Sets.newHashSet();
+    for (FileStatus fileStatus : FileListUtils.listFilesRecursively(fs, paths)) {
+      if (fileStatus.getPath().toString().contains(MRCompactor.COMPACTION_RENAME_SOURCE_DIR_SUFFIX)) {
+        renamedDirs.add(fileStatus.getPath().getParent());
+      }
+    }
+
+    return renamedDirs;
+  }
+
+    /**
+     * Get all the unrenamed directories from {@link Dataset#inputPaths}
+     * They are deepest level directories which doesn't have a suffix
+     * {@link MRCompactor#COMPACTION_RENAME_SOURCE_DIR_SUFFIX} append.
+     */
+  public static Set<Path> getInputDeepestLevelUnrenamedDirsWithFileExistence (FileSystem fs, Set<Path> paths) throws IOException {
+    Set<Path> unrenamed = Sets.newHashSet();
+    for (FileStatus fileStatus : FileListUtils.listFilesRecursively(fs, paths)) {
+      if (!fileStatus.getPath().toString().contains(MRCompactor.COMPACTION_RENAME_SOURCE_DIR_SUFFIX)) {
```

Comment:
I use two method because they are not mutual exclusive when file existence is taken into consideration.